pkg/status/reporter.go

```diff
@@ -0,0 +1,209 @@
+	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+)
+
+// Interface for a component to populate its status to node status resource.
+type statusPopulator interface {
+	Populate(status *apiv3.CalicoNodeStatus) error
+	Show()
+}
+
+// reporter contains all the data/method about reporting back node status based on a single node status resource.
+// Each reporter has a goroutine which constantly reads node status and updates node status resource.
+type reporter struct {
+	// The name of the node status resource.
+	name string
+
+	// Internal client to operate on node status resource.
+	client client.Interface
+
+	// buffered channel to receive updates for the resource.
+	ch chan *apiv3.CalicoNodeStatus
+
+	// status holds latest version of node status resource.
+	status *apiv3.CalicoNodeStatus
+
+	// Interval and Time ticker that node status should be reported.
+	interval int
+	ticker   *time.Ticker
+
+	// populators
+	populators map[BirdConnType]map[apiv3.NodeStatusClassType]statusPopulator
+
+	// channel to indicate this reporter is not needed anymore.
+	// It should start termination process.
+	done chan struct{}
+
+	// channel to indicate this reporter is terminated.
+	term chan struct{}
+}
+
+// newReporter creates a reporter and start running a goroutine handling resource update.
+func newReporter(name string, client client.Interface, intervalInSeconds int, populators map[BirdConnType]map[apiv3.NodeStatusClassType]statusPopulator) *reporter {
+	r := &reporter{
+		name:       name,
+		client:     client,
+		ch:         make(chan *apiv3.CalicoNodeStatus, 10),
+		populators: populators,
+		done:       make(chan struct{}),
+	}
+
+	r.checkAndUpdateTicker(intervalInSeconds)
+
+	go r.run()
+	return r
+}
+
+// Check and set new ticker for the reporter.
+// Make sure stop the old one first to GC old ticker.
+func (r reporter) checkAndUpdateTicker(interval int) {
+	if r.interval == interval {
+		// no update needed.
+		return
+	}
+	if r.ticker != nil {
+		r.ticker.Stop()
+	}
+	r.interval = interval
+	r.ticker = time.NewTicker(time.Duration(interval) * time.Second)
+}
+
+// Cleanup resources owned by this reporter.
+func (r reporter) cleanup() {
+	r.ticker.Stop()
+}
+
+// KillAndWait sends done signal to reporter goroutine and wait until the
+// goroutine of the reporter terminated.
+func (r reporter) KillAndWait() {
+	r.done <- struct{}{}
+	<-r.term
+}
+
+// Called when the caller needs to send a new version of request.
+func (r reporter) RequestUpdate(request *apiv3.CalicoNodeStatus) {
+	r.ch <- request
+}
+
+// run is the main reporting loop, it loops until done.
+func (r reporter) run() {
+	runImmediately := make(chan struct{})
+
+	for {
+		select {
+		case latest := <-r.ch:
+			// Received an update of node status resource.
+			if latest.Name != r.name {
+				log.Warningf("node status reporter (%s) receive request with different name (%s), ignore it", r.name, latest.Name)
```

Comment:
This is really a bug in the code if we hit this, right? We can probably make this louder (error level).

I'd also recommend simplifying the conditional logic here by putting a `continue` in this branch and removing the `else`, which makes it more clear that this is not the mainline code path.